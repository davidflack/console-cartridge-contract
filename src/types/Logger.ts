export interface Logger {
  (...args: any[]): void;
}
