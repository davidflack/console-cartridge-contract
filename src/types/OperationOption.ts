export interface OperationOption<InputType, OutputType> {
  name: string,
  enabled: boolean,
  option: (inputType: InputType) => OutputType | Promise<OutputType>
}
