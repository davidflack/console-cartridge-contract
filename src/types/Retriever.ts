export interface Retriever<I, O> {
  (inputType: I): Promise<O>
}
