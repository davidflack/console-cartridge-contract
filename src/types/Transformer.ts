export interface TransformerWithoutTracing<T, O> {
  (inputType: T): O;
}
export interface TransformerWithTracing<T, O> {
  (id: string): TransformerWithoutTracing<T, O>; // TODO: is this the best approach here?
}

export type Transformer<T, O> =
  | TransformerWithoutTracing<T, O>
  | TransformerWithTracing<T, O>;
