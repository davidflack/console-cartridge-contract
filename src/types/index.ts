export * from './Cartridge';
export * from './Logger';
export * from './Operation';
export * from './OperationOption';
export * from './Retriever';
export * from './Transformer';
