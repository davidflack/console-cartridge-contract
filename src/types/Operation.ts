import { Retriever } from './Retriever';
import { Transformer } from './Transformer';

interface IOWithoutTransformation<
  InputType,
  OutputType,
> {
  operation: Retriever<InputType, OutputType>,
}

interface IOWithTransformation<
  InputType,
  OutputType,
  RequestType,
  ResponseType
> extends IOWithoutTransformation<RequestType, ResponseType> {
  transform: {
    request: Transformer<InputType, RequestType>,
    response: Transformer<ResponseType, OutputType>
  },
}

type IO<
  InputType,
  OutputType,
  RequestType = void,
  ResponseType = void
> = IOWithTransformation<InputType, OutputType, RequestType, ResponseType> | IOWithoutTransformation<InputType, OutputType>

export interface Operation<
  InputType,
  RequestType,
  ResponseType,
  OutputType
> {
  name: String,
  steps: IO<InputType, OutputType, RequestType, ResponseType>[],
}
