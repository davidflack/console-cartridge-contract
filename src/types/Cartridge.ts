import { Logger } from './Logger';
import { Operation } from './Operation';

export interface Cartridge {
  name: string;
  operations: Operation<unknown, unknown, unknown, unknown>[];
  logger: Logger;
}
