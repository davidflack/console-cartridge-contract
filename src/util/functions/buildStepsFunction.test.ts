import { get, join, map, toUpper } from 'lodash';

import { Operation } from '@console-cartridge-contract/types';

import {
  buildStepFunctionsFromOperation,
  buildStepsFunction,
  mapSeries,
} from './buildStepsFunction';

const transformerlessTestOperation: Operation<string, any, any, string> = {
  name: 'transformerlessOperation',
  steps: [
    {
      operation: async (n: string) => toUpper(n),
    },
  ],
};

enum Weapon {
  NUNCHUCKS = 'NUNCHUCKS',
  SAI = 'SAI',
  KATANA = 'KATANA',
  BO = 'BO',
}

enum Color {
  BLUE = 'BLUE',
  PURPLE = 'PURPLE',
  RED = 'RED',
  ORANGE = 'ORANGE',
}

enum TurtleName {
  LEONARDO = 'LEONARDO',
  DONATELLO = 'DONATELLO',
  RAPHAEL = 'RAPHAEL',
  MICHAELANGELO = 'MICHAELANGELO',
}

interface NinjaTurtle {
  name: string;
  notes?: string;
  weapon?: Weapon;
  color?: Color;
}

interface ExampleIOResponse {
  data: any;
}

/**
 * Mock a call to a system that finds a Ninja Turtle's weapon.
 *
 * @param n
 * @returns
 */
const getWeaponFromAPI = async (n: string) => {
  const ninjaTurtle = JSON.parse(n);
  const name = get(ninjaTurtle, 'name');

  const buildResponse = (weapon: Weapon) => ({
    data: JSON.stringify({
      ...ninjaTurtle,
      weapon,
    }),
  });

  switch (toUpper(name)) {
    case TurtleName.DONATELLO:
      return buildResponse(Weapon.BO);
    case TurtleName.LEONARDO:
      return buildResponse(Weapon.KATANA);
    case TurtleName.MICHAELANGELO:
      return buildResponse(Weapon.NUNCHUCKS);
    case TurtleName.RAPHAEL:
      return buildResponse(Weapon.SAI);
    default:
      throw new Error(`${name} is not a Ninja Turtle!`);
  }
};

/**
 * Mock a call to a system that finds Ninja Turtle's colors.
 *
 * @param n
 * @returns
 */
const getVocalsFromAPI = async (n: string) => {
  const ninjaTurtle = JSON.parse(n);
  const name = get(ninjaTurtle, 'name');

  const buildResponse = (color: Color) => ({
    data: JSON.stringify({
      ...ninjaTurtle,
      color,
    }),
  });

  switch (toUpper(name)) {
    case TurtleName.DONATELLO:
      return buildResponse(Color.PURPLE);
    case TurtleName.LEONARDO:
      return buildResponse(Color.BLUE);
    case TurtleName.MICHAELANGELO:
      return buildResponse(Color.ORANGE);
    case TurtleName.RAPHAEL:
      return buildResponse(Color.RED);
    default:
      throw new Error(`${name} is not a Ninja Turtle!`);
  }
};

const getNinjaTurtleInfo: Operation<NinjaTurtle, any, any, NinjaTurtle> = {
  name: 'getNinjaTurtleInfo',
  steps: [
    {
      operation: getWeaponFromAPI,
      transform: {
        request: (ninjaTurtle: NinjaTurtle) => JSON.stringify(ninjaTurtle),
        response: ({ data }: ExampleIOResponse) => JSON.parse(data),
      },
    },
    {
      operation: getVocalsFromAPI,
      transform: {
        request: (ninjaTurtle: NinjaTurtle) => JSON.stringify(ninjaTurtle),
        response: ({ data }: ExampleIOResponse) => JSON.parse(data),
      },
    },
  ],
};

describe(buildStepFunctionsFromOperation, () => {
  it('builds step functions without a transformer', async () => {
    expect(
      buildStepFunctionsFromOperation(transformerlessTestOperation),
    ).toMatchSnapshot();
  });
  it('builds step functions with transformers', async () => {
    expect(
      buildStepFunctionsFromOperation(getNinjaTurtleInfo),
    ).toMatchSnapshot();
  });
});

describe(mapSeries, () => {
  it('executes functions in series', async () => {
    const mockStepFunctions = [
      async (n) => `${n}_1`,
      async (n) => `${n}_2`,
      async (n) => `${n}_3`,
    ];
    expect(await mapSeries(mockStepFunctions, 'input')).toEqual('input_1_2_3');
  });
  it('executes non-heterogeneous functions in series', async () => {
    const mockStepFunctions = [
      async (n: string[], toUpperCase: boolean) =>
        toUpperCase ? map(n, toUpper) : n,
      async (n: string[]) => join(n, ','),
    ];
    expect(
      await mapSeries(
        mockStepFunctions,
        ['Donatello', 'Leonardo', 'Michaelangelo', 'Rafael'],
        true,
      ),
    ).toEqual('DONATELLO,LEONARDO,MICHAELANGELO,RAFAEL');
  });
});

describe(buildStepsFunction, () => {
  it('compiles multiple steps without transformers into one function', async () => {
    const stepsFunction = buildStepsFunction(transformerlessTestOperation);
    expect(await stepsFunction('foo')).toEqual('FOO');
  });
  it('compiles multiple steps with transformers into one function', async () => {
    const stepsFunction = buildStepsFunction(getNinjaTurtleInfo);
    expect(
      await stepsFunction({
        name: 'Donatello',
      }),
    ).toMatchSnapshot();
  });
  it('logs all of the steps along the way', async () => {
    const log = jest.fn();
    const stepsFunction = buildStepsFunction(getNinjaTurtleInfo, log);
    await stepsFunction({
      name: 'Michaelangelo',
    });
    expect(log.mock.calls).toMatchSnapshot();
  });
});
