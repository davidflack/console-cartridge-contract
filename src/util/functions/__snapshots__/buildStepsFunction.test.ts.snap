// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildStepFunctionsFromOperation builds step functions with transformers 1`] = `
Array [
  [Function],
  [Function],
]
`;

exports[`buildStepFunctionsFromOperation builds step functions without a transformer 1`] = `
Array [
  [Function],
]
`;

exports[`buildStepsFunction compiles multiple steps with transformers into one function 1`] = `
Object {
  "color": "PURPLE",
  "name": "Donatello",
  "weapon": "BO",
}
`;

exports[`buildStepsFunction logs all of the steps along the way 1`] = `
Array [
  Array [
    "getNinjaTurtleInfo-step-1-request-transformer",
    "input",
    Object {
      "name": "Michaelangelo",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-1-request-transformer",
    "output",
    "{\\"name\\":\\"Michaelangelo\\"}",
  ],
  Array [
    "getNinjaTurtleInfo-step-1-operation",
    "input",
    Object {
      "name": "Michaelangelo",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-1-operation",
    "output",
    Object {
      "data": "{\\"name\\":\\"Michaelangelo\\",\\"weapon\\":\\"NUNCHUCKS\\"}",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-1-response-transformer",
    "input",
    Object {
      "data": "{\\"name\\":\\"Michaelangelo\\",\\"weapon\\":\\"NUNCHUCKS\\"}",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-1-response-transformer",
    "output",
    Object {
      "name": "Michaelangelo",
      "weapon": "NUNCHUCKS",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-2-request-transformer",
    "input",
    Object {
      "name": "Michaelangelo",
      "weapon": "NUNCHUCKS",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-2-request-transformer",
    "output",
    "{\\"name\\":\\"Michaelangelo\\",\\"weapon\\":\\"NUNCHUCKS\\"}",
  ],
  Array [
    "getNinjaTurtleInfo-step-2-operation",
    "input",
    Object {
      "name": "Michaelangelo",
      "weapon": "NUNCHUCKS",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-2-operation",
    "output",
    Object {
      "data": "{\\"name\\":\\"Michaelangelo\\",\\"weapon\\":\\"NUNCHUCKS\\",\\"color\\":\\"ORANGE\\"}",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-2-response-transformer",
    "input",
    Object {
      "data": "{\\"name\\":\\"Michaelangelo\\",\\"weapon\\":\\"NUNCHUCKS\\",\\"color\\":\\"ORANGE\\"}",
    },
  ],
  Array [
    "getNinjaTurtleInfo-step-2-response-transformer",
    "output",
    Object {
      "color": "ORANGE",
      "name": "Michaelangelo",
      "weapon": "NUNCHUCKS",
    },
  ],
]
`;
